### PR TITLE
[dev] [DeepSeek-v4] Part 3: MTP support with mHC and new mHC contract

### PR DESCRIPTION
**Pull request by hxbai**
# What does this PR do ?
We will create several PRs to functionally support DeepSeek-v4 training. This is the third one.

Add DeepSeek-v4 MTP support with mHC and new mHC contract.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/transformer_block.py [hunk 1]
```diff
@@ -381,6 +385,17 @@ def build_layer(layer_spec, layer_number):
...
+                hc_mult = self.config.num_residual_streams
+                hc_dim = self.config.hidden_size * hc_mult
+                self.hc_head_fn = nn.Parameter(torch.randn(hc_mult, hc_dim))
+                self.hc_head_base = nn.Parameter(torch.zeros(hc_mult))
+                self.hc_head_scale = nn.Parameter(torch.ones(1))
+                nn.init.xavier_uniform_(self.hc_head_fn)
+                if self.config.sequence_parallel:
+                    setattr(self.hc_head_fn, 'sequence_parallel', True)
+                    setattr(self.hc_head_base, 'sequence_parallel', True)
+                    setattr(self.hc_head_scale, 'sequence_parallel', True)
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/transformer_block.py:402]: @hxbai this one make sense and I think without wrapping those are not handled correctly by dist ckpting. These parameters are meant to be loaded on restart, right?

File: megatron/core/transformer/transformer_block.py [hunk 2]
```diff
@@ -949,6 +977,11 @@ def forward(
         if len(extract_layer_indices) > 0:
             return hidden_states, intermediate_hidden_states
 
+        # When mHC + MTP, return both contracted [s,b,h] (for lm_head) and
+        # pre-contraction multi-stream [s,b,n*h] (for MTP input).
+        if mhc_multistream is not None:
+            return hidden_states, mhc_multistream
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/transformer_block.py:988]: will migrate to hybrid_model in a separate PR

**Review by FDecaYed — APPROVED** [review]

File: megatron/core/transformer/transformer_block.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/transformer_block.py:402]: fixed

File: megatron/core/transformer/transformer_block.py [hunk 3]
```diff
@@ -926,10 +941,23 @@ def forward(
...
+                mhc_multistream = hidden_states
+            # [s, b, n*C] -> [s, b, C]
+            hidden_states = learned_output_contract(
+                hidden_states,
+                self.hc_head_fn,
+                self.hc_head_base,
+                self.hc_head_scale,
+                self.config.num_residual_streams,
+                self.config.layernorm_epsilon,
+            )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/transformer_block.py:965]: This is intentional. mhc is only used in DSv4 and DSv4 changed the structure.

File: megatron/core/transformer/transformer_block.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/transformer_block.py:988]: hybrid model will be handled separately later. Leave it now.

File: megatron/core/transformer/hyper_connection.py [hunk 4]
```diff
@@ -91,6 +92,23 @@ def native_proj_rms(x: Tensor, weight: Tensor, eps: float = 1e-6) -> Tuple[Tenso
...
+    dtype = hidden_states.dtype
+    hidden_states = hidden_states.to(torch.float32)
+    head_fn = head_fn.to(torch.float32)
+    base = base.to(torch.float32)
+    scale = scale.to(torch.float32)
+    rsqrt = torch.rsqrt(hidden_states.square().mean(-1, keepdim=True) + eps)
+    mixes = F.linear(hidden_states, head_fn) * rsqrt
+    pre = torch.sigmoid(mixes * scale + base) + 1e-6
+    y = torch.sum(pre.unsqueeze(-1) * hidden_states.view(*hidden_states.shape[:-1], n, -1), dim=-2)
+    return y.to(dtype)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/hyper_connection.py:109]: replaced with eps

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 5]
```diff
@@ -120,6 +120,10 @@ def _apply_rope(
             ), "Fused MLA RoPE apply is not imported successfully"
         else:
             rotary_pos_emb, mscale = rotary_pos_emb_module(total_seq_len, packed_seq=False)
+            # DSv4 reference (DS-Inf) RoPE is pure rotation (norm-preserving). Yarn's
+            # concentration factor (mscale) is NOT part of the DSv4 model contract --
+            # the model relies on Q/KV RMS-norm + unit-magnitude rotation. Force 1.0.
+            mscale = 1.0
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:126]: This is the special behavior of DSv4 and no need to put it into a standalone position.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/726ab33851a565e9372b5e0cc042f1f4744e12ff

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/c44161446ad52f3d38e61136fb2e51c0a6e1cd22

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/02b7730c2a52607bc8020b56d253d7ea143e1a35

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/43177bb60d4c483881dd9598110049a0a8d1c852

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/8c31b4934898f018dbb36be5ec0737a78c82de42

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/a78d1660200223bc2a81e713c129ef4c9b91f96f

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/e0514d83c1886fc0e7d55716b8dd51b6f55aec85

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/93757b4a3721cf26723f2b0b5fe5a8349b668ae3

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/ae28a198c0d5b6145b335ad1cba5ae5decb4ec21

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/8fd14d5263c58e0f6124fda37c493cb80e3c6808

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/be7dd1bfb10dd7f4976d5b7fcad394724b53c6ad

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4518/commits/60ff377d0eff380231e27332030c3a322edb03d6
